recipe/build-grpcio.bat

```diff
@@ -1,5 +1,6 @@
 if "%ARCH%" == "32" (set PLATFORM=x86) else (set PLATFORM=x64)
 
+set "EXTRA_ENV_COMPILE_ARGS=/std:c++14"
```

Comment:
You cannot have mismatched standard versions if compiling against abseil. I mean, you can, and if you're very lucky you won't be affected by any of the divergences, but it's not a good idea

Reply:
ok, well maybe we just leap over the latest abseil to the next one. They seem to have disable c++11 there, so we can move on from something that more than 10 years old.

Reply:
> ok, well maybe we just leap over the latest abseil to the next one. They seem to have disable c++11 there, so we can move on from something that more than 10 years old.

The issue is that abseil - uniquely, to my knowledge at least - changes its API/ABI based on the C++ standard. For that reason it makes no sense to build (only) dynamic libs, because we can never impose a homogeneous C++ for all feedstocks. The only real solution here is to additionally build static libs for all C++ standards and let each feedstock use the matching C++ standard, if they cannot used the C++ version used for the shared lib

Reply:
So the issue is not solved by abseil removing the C++11 option. It just shifts upwards to C++14/C++17/C++20. Hence https://github.com/conda-forge/abseil-cpp-feedstock/pull/35

Reply:
Lets see how much we can force c++11 on windows by patching out the check.

Reply:
i noticed that you called out the arrow team for c++11->14 transition. So it seems that some are still interested in holding back c++11 for windows at least.

Reply:
> i noticed that you called out the arrow team for c++11->14 transition. So it seems that some are still interested in holding back c++11 for windows at least.

They're interested in [moving](https://issues.apache.org/jira/browse/ARROW-17110). Again the sustainable solution would be to let people choose the version they need (statically).

I rebased https://github.com/conda-forge/abseil-cpp-feedstock/pull/35, but I'm having silly errors about jinja-vs-conda-vs-mamba. The concept is easy to express, but I'd like to be able to do it without copy-pasting the same output 3x. Help would be appreciated.

Reply:
I guess i don't really agree that static libraries are sustainable.

In terms of syntax, I might go with build variants specified in the `conda_build_config.yaml` and omit the `libabseil` and `abseil-cpp` from the ones you deem "inferior".

It seems that the build errors are the same for the build where i forced c++11 and c++14. So I don't think we have identified the source of the problem here.

Reply:
The https://github.com/conda-forge/ffmpeg-feedstock/blob/main/recipe/conda_build_config.yaml which I believe you have reviewed should be a good example of what I mean.

Reply:
> I guess i don't really agree that static libraries are sustainable.

Sure, in general that's true, but not for abseil. I'm really adamant about that, because it's simply not a sustainable way to run an ecosystem across extremely diverse packages by dictating a uniform CXX standard to compile with (plus the debugging pain when you don't know is massive - chasing mysterious linker errors etc.)

> In terms of syntax, I might go with build variants specified in the `conda_build_config.yaml` and omit the `libabseil` and `abseil-cpp` from the ones you deem "inferior".

I tried that in many ways, it always failed. I pushed my old branch so you can see all that I've tried: https://github.com/h-vetinari/abseil-cpp-feedstock/tree/std_old

> The https://github.com/conda-forge/ffmpeg-feedstock/blob/main/recipe/conda_build_config.yaml which I believe you have reviewed should be a good example of what I mean.

It might be a problem that occurs only with multi-output builds, which have somewhat different behaviour sometimes.

Reply:
I'm fairly convinced I am now building with c++11 as the standard and still seeing the errors. what is the standard of abseil? it seems to be undefined in the present build process.

Reply:
https://docs.microsoft.com/en-us/answers/questions/182681/what-default-c-version-set-in-vs-2017.html point toward c++14 on vs2017

Reply:
> I'm fairly convinced I am now building with c++11 as the standard and still seeing the errors. what is the standard of abseil? it seems to be undefined in the present build process.

You're building with C++11, but overriding the minimum set by grpc. If they use any of the post-C++11 ABI the build cannot work.

Reply:
> > I guess i don't really agree that static libraries are sustainable.
> 
> Sure, in general that's true, but not for abseil. I'm really adamant about that, because it's simply not a sustainable way to run an ecosystem across extremely diverse packages by dictating a uniform CXX standard to compile with (plus the debugging pain when you don't know is massive - chasing mysterious linker errors etc.) - and obviously we cannot have the same shared lib in two different flavours in the same environment.

It was late yesterday so some more context: it's not true for abseil _precisely_ because the ABI is sensitive to the standards - with all other libs you just need to worry about the version, set the right level of run_export, and things work. But if the same version has a different ABI based on the CXX standard, you're screwed.

Also, I'm proposing that the default would still be dynamic builds for those feedstocks that can handle C++17, but provide an escape hatch for those that cannot. And that escape hatch _cannot_ be dynamic, because you cannot have the same version of abseil with two different ABIs as a runtime dependence.

Reply:
So IMO: let's get some static builds sorted (windows is always static anyway, or at least the shared lib is incomplete), and try this again with C++14. Thanks for https://github.com/conda-forge/abseil-cpp-feedstock/pull/38, that was a really obvious-in-hindsight mistake on my part, but I couldn't see it. I'll look at it in more detail later.